[3.7] bpo-17288: Prevent jumps from 'return' and 'exception' trace events

xdegaye opened the pull request:
<!-- issue-number: bpo-17288 -->
https://bugs.python.org/issue17288
<!-- /issue-number -->

xdegaye committed:
bpo-17288: Prevent jumps from 'return' and 'exception' trace events

serhiy-storchaka commented on Lib/test/test_sys_settrace.py:
```diff
@@ -1076,6 +1101,22 @@ class fake_function:
         sys.settrace(None)
         self.compare_jump_output([2, 3, 2, 3, 4], namespace["output"])
 
+    def test_no_jump_from_call(self):
```
Could tests use the `jump_test()` decorator?

serhiy-storchaka commented on Objects/frameobject.c:
```diff
@@ -94,13 +97,34 @@ frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno)
+     * Forbidding jumps from the 'call' event of a new frame is a side effect
+     * of allowing to set f_lineno only from trace functions. */
+    if (f->f_lasti == -1)
+    {
```
Please left this brace on the same line as `if` to conform PEP 7.

serhiy-storchaka commented on Objects/frameobject.c:
```diff
@@ -159,6 +183,16 @@ frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno)
+    /* The trace function is called with a 'return' trace event after the
+     * execution of a yield statement. */
+    assert(f->f_lasti != -1);
+    if (code[f->f_lasti] == YIELD_VALUE || code[f->f_lasti] == YIELD_FROM) {
```
Is there a need to check for RETURN_VALUE as in your original patch?

xdegaye replied:
The only function called by ``_PyEval_EvalFrameDefault()`` that may allow a jump to another line is ``maybe_call_line_trace()`` because ``stack_pointer`` is saved on ``f->f_stacktop`` before the call and there is a ``JUMPTO(f->f_lasti)`` after the call. All the other trace functions called within ``_PyEval_EvalFrameDefault()`` are called with ``f->f_stacktop == NULL`` except after the handling of the ``YIELD_FROM`` and ``YIELD_VALUE`` opcodes. So by forbidding jumps in ``frame_setlineno()`` when ``f->f_stacktop == NULL`` we handle all the cases, including the ``RETURN_VALUE`` opcode, except the yield statements. The check for ``RETURN_VALUE`` in my original patch was overkill.

xdegaye committed:
The new tests use now the jump_test() decorator

Fix also a bug in the previous implementation of the PR where
jumpFrom/jumpTo were systematically incremented by 1 by the
jump_test() decorator without taking into account that
JumpTracer allows now jumpFrom/jumpTo from within nested
functions.

xdegaye committed:
JumpTestCase line numbers are relative to the first line of the testing function

serhiy-storchaka reviewed (approved):
Thank you @xdegaye!

I have removed comments for brevity in the example. Could you please restore the old comment about non-integer jumpTo? Or maybe even add new comments to non-trivial code?

xdegaye committed:
Restore previous comments
